src/backend/distributed/utils/citus_depended_object.c

```diff
@@ -0,0 +1,292 @@
+/*
+ * citus_depended_object.c
+ *
+ * Implements exposed functions related to hiding citus depended objects.
+ *
+ * Copyright (c) Citus Data, Inc.
+ */
+
+#include "postgres.h"
+#include "miscadmin.h"
+
+#include "catalog/namespace.h"
+#include "catalog/pg_aggregate.h"
+#include "catalog/pg_am.h"
+#include "catalog/pg_attribute.h"
+#include "catalog/pg_attrdef.h"
+#include "catalog/pg_constraint.h"
+#include "catalog/pg_class.h"
+#include "catalog/pg_depend.h"
+#include "catalog/pg_enum.h"
+#include "catalog/pg_event_trigger.h"
+#include "catalog/pg_language.h"
+#include "catalog/pg_namespace.h"
+#include "catalog/pg_proc.h"
+#include "catalog/pg_rewrite.h"
+#include "catalog/pg_sequence.h"
+#include "catalog/pg_statistic.h"
+#include "catalog/pg_trigger.h"
+#include "catalog/pg_ts_config.h"
+#include "catalog/pg_ts_dict.h"
+#include "catalog/pg_ts_template.h"
+#include "catalog/pg_type.h"
+#include "distributed/citus_depended_object.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/listutils.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/parsenodes.h"
+#include "utils/lsyscache.h"
+
+/*
+ * GUC hides any objects, which depends on citus extension, from pg meta class queries,
+ * it is intended to be used in vanilla tests to not break postgres test logs
+ */
+bool HideCitusDependentObjects = false;
+
+static Node * CreateCitusDependentObjectExpr(int pgMetaTableVarno, int pgMetaTableOid);
+static List * GetCitusDependedObjectArgs(int pgMetaTableVarno, int pgMetaTableOid);
+
+/*
+ * IsPgLocksTable returns true if RTE is pg_locks table.
+ */
+bool
+IsPgLocksTable(RangeTblEntry *rte)
+{
+	Oid pgLocksId = get_relname_relid("pg_locks", get_namespace_oid("pg_catalog", false));
+	return rte->relid == pgLocksId;
+}
+
+
+/*
+ * SetLocalHideCitusDependentObjects sets the GUC HideCitusDependentObjects to
+ * given value locally.
+ */
+void
+SetLocalHideCitusDependentObjects(bool state)
+{
+	const char *stateVal = (state) ? "true" : "false";
+	set_config_option("citus.hide_citus_dependent_objects", stateVal,
+					  (superuser() ? PGC_SUSET : PGC_USERSET), PGC_S_SESSION,
+					  GUC_ACTION_LOCAL, true, 0, false);
+}
+
+
+/*
+ * HideCitusDependentObjectsFromPgMetaTable adds a NOT is_citus_depended_object(oid, oid) expr
+ * to the quals of meta class RTEs that we are interested in.
+ */
+bool
+HideCitusDependentObjectsFromPgMetaTable(Node *node, void *context)
+{
+	if (!CitusHasBeenLoaded())
```

Comment:
minor: We can probably merge all 3 `if`s into one for ease of read

Reply:
Or, maybe even `Assert(HideCitusDependentObjects == true)` such that we never call this accidentally with other values